File: src/shared/platform/tauri/feature-ports/projects.ts
import { invoke } from "@tauri-apps/api/core";
import {
  bindKitSchema,
  credentialBindingSchema,
  credentialsSchema,
  customerSchema,
  customersSchema,
  dependencySnapshotSchema,
  mutationSchema,
  projectContextSchema,
  projectIdSchema,
  projectSchema,
  projectsSchema,
  resourceKindSchema,
  resourcesSchema,
} from "../../../../domain/projects";
import type { ProjectsPort } from "../../../features/projects";

export function createProjectsPort(): ProjectsPort {
  return {
    listCustomers: async () =>
      customersSchema.parse(await invoke<unknown>("projects_list_customers")),
    createCustomer: async (name) =>
      customerSchema.parse(
        await invoke<unknown>("projects_create_customer", { name }),
      ),
    updateCustomer: async (customerId, expectedRevision, name, archived) =>
      customerSchema.parse(
        await invoke<unknown>("projects_update_customer", {
          customerId: projectIdSchema.parse(customerId),
          expectedRevision,
          name,
          archived,
        }),
      ),
    list: async () =>
      projectsSchema.parse(await invoke<unknown>("projects_list")),
    get: async (projectId) =>
      projectSchema.parse(
        await invoke<unknown>("projects_get", {
          projectId: projectIdSchema.parse(projectId),
        }),
      ),
    create: async (customerId, name) =>
      projectSchema.parse(
        await invoke<unknown>("projects_create", {
          customerId: projectIdSchema.parse(customerId),
          name,
        }),
      ),
    update: async (request, name, archived) =>
      projectSchema.parse(
        await invoke<unknown>("projects_update", {
          request: mutationSchema.parse(request),
          name,
          archived,
        }),
      ),
    resourceOptions: async () =>
      resourcesSchema.parse(await invoke<unknown>("projects_resource_options")),
    credentialOptions: async () =>
      credentialsSchema.parse(
        await invoke<unknown>("projects_credential_options"),
      ),
    bindResource: async (request, resource, model) =>
      projectSchema.parse(
        await invoke<unknown>("projects_bind_resource", {
          request: mutationSchema.parse(request),
          kind: resourceKindSchema.parse(resource.kind),
          agentId: resource.agentId,
          rawId: resource.rawId,
          model,
        }),
      ),
    removeResource: async (request, resource) =>
      projectSchema.parse(
        await invoke<unknown>("projects_remove_resource", {
          request: mutationSchema.parse(request),
          kind: resourceKindSchema.parse(resource.kind),
          agentId: resource.agentId,
          rawId: resource.rawId,
        }),
      ),
    bindCredential: async (request, credential) => {
      const c = credentialBindingSchema.parse({
        credentialId: credential.credentialId,
        purpose: credential.purpose,
        consumer: credential.consumer,
        pinnedGeneration: 0,
      });
      return projectSchema.parse(
        await invoke<unknown>("projects_bind_credential", {
          request: mutationSchema.parse(request),
          credentialId: c.credentialId,
          purpose: c.purpose,
          consumer: c.consumer,
        }),
      );
    },
    removeCredential: async (request, credentialId) =>
      projectSchema.parse(
        await invoke<unknown>("projects_remove_credential", {
          request: mutationSchema.parse(request),
          credentialId,
        }),
      ),
    getContext: async (projectId) =>
      projectContextSchema.parse(
        await invoke<unknown>("projects_get_context", {
          projectId: projectIdSchema.parse(projectId),
        }),
      ),
    writeContext: async (request, content) =>
      projectContextSchema.parse(
        await invoke<unknown>("projects_write_context", {
          request: mutationSchema.parse(request),
          content,
        }),
      ),
    prepareCodex: async (request) =>
      projectContextSchema.parse(
        await invoke<unknown>("projects_prepare_codex", {
          request: mutationSchema.parse(request),
        }),
      ),
    bindDeliveryKit: async (request) =>
      projectSchema.parse(
        await invoke<unknown>("projects_bind_delivery_kit", {
          request: bindKitSchema.parse(request),
        }),
      ),
    dependencySnapshot: async (projectId) =>
      dependencySnapshotSchema.parse(
        await invoke<unknown>("projects_dependency_snapshot", {
          projectId: projectIdSchema.parse(projectId),
        }),
      ),
  };
}

File: src-tauri/src/commands/projects.rs
use crate::services::projects::{domain::*, ProjectsService};
use tauri::State;

fn public_error(e: crate::error::AppError) -> String {
    if let crate::error::AppError::InvalidInput(code) = &e {
        if code.starts_with("projects_") {
            return code.clone();
        }
    }
    "projects_unavailable".into()
}
#[tauri::command]
pub(crate) fn projects_list_customers(
    state: State<'_, std::sync::Arc<ProjectsService>>,
) -> Result<Vec<Customer>, String> {
    state.customers().map_err(public_error)
}
#[tauri::command]
pub(crate) fn projects_create_customer(
    state: State<'_, std::sync::Arc<ProjectsService>>,
    name: String,
) -> Result<Customer, String> {
    state.create_customer(&name).map_err(public_error)
}
#[tauri::command]
pub(crate) fn projects_update_customer(
    state: State<'_, std::sync::Arc<ProjectsService>>,
    customer_id: String,
    expected_revision: i64,
    name: String,
    archived: bool,
) -> Result<Customer, String> {
    state
        .update_customer(&customer_id, expected_revision, &name, archived)
        .map_err(public_error)
}
#[tauri::command]
pub(crate) fn projects_list(
    state: State<'_, std::sync::Arc<ProjectsService>>,
) -> Result<Vec<Project>, String> {
    state.list().map_err(public_error)
}
#[tauri::command]
pub(crate) fn projects_get(
    state: State<'_, std::sync::Arc<ProjectsService>>,
    project_id: String,
) -> Result<Project, String> {
    state.get(&project_id).map_err(public_error)
}
#[tauri::command]
pub(crate) fn projects_create(
    state: State<'_, std::sync::Arc<ProjectsService>>,
    customer_id: String,
    name: String,
) -> Result<Project, String> {
    state.create(&customer_id, &name).map_err(public_error)
}
#[tauri::command]
pub(crate) fn projects_update(
    state: State<'_, std::sync::Arc<ProjectsService>>,
    request: ProjectMutation,
    name: String,
    archived: bool,
) -> Result<Project, String> {
    state
        .update(&request, &name, archived)
        .map_err(public_error)
}
#[tauri::command]
pub(crate) fn projects_resource_options(
    state: State<'_, std::sync::Arc<ProjectsService>>,
) -> Result<Vec<ResourceOption>, String> {
    state.resource_options().map_err(public_error)
}
#[tauri::command]
pub(crate) fn projects_credential_options(
    state: State<'_, std::sync::Arc<ProjectsService>>,
) -> Result<Vec<CredentialOption>, String> {
    state.credential_options().map_err(public_error)
}
#[tauri::command]
pub(crate) fn projects_bind_resource(
    state: State<'_, std::sync::Arc<ProjectsService>>,
    request: ProjectMutation,
    kind: ResourceKind,
    agent_id: String,
    raw_id: String,
    model: Option<String>,
) -> Result<Project, String> {
    state
        .bind_resource(&request, kind, &agent_id, &raw_id, model)
        .map_err(public_error)
}
#[tauri::command]
pub(crate) fn projects_remove_resource(
    state: State<'_, std::sync::Arc<ProjectsService>>,
    request: ProjectMutation,
    kind: ResourceKind,
    agent_id: String,
    raw_id: String,
) -> Result<Project, String> {
    state
        .remove_resource(&request, kind, &agent_id, &raw_id)
        .map_err(public_error)
}
#[tauri::command]
pub(crate) fn projects_bind_credential(
    state: State<'_, std::sync::Arc<ProjectsService>>,
    request: ProjectMutation,
    credential_id: String,
    purpose: String,
    consumer: String,
) -> Result<Project, String> {
    state
        .bind_credential(&request, &credential_id, &purpose, &consumer)
        .map_err(public_error)
}
#[tauri::command]
pub(crate) fn projects_remove_credential(
    state: State<'_, std::sync::Arc<ProjectsService>>,
    request: ProjectMutation,
    credential_id: String,
) -> Result<Project, String> {
    state
        .remove_credential(&request, &credential_id)
        .map_err(public_error)
}
#[tauri::command]
pub(crate) fn projects_get_context(
    state: State<'_, std::sync::Arc<ProjectsService>>,
    project_id: String,
) -> Result<ProjectContext, String> {
    state.context(&project_id).map_err(public_error)
}
#[tauri::command]
pub(crate) fn projects_write_context(
    state: State<'_, std::sync::Arc<ProjectsService>>,
    request: ProjectMutation,
    content: String,
    recover: Option<bool>,
) -> Result<ProjectContext, String> {
    if recover.unwrap_or(false) {
        state.write_context_with_recovery(&request, &content, true)
    } else {
        state.write_context(&request, &content)
    }
    .map_err(public_error)
}
#[tauri::command]
pub(crate) fn projects_bind_delivery_kit(
    state: State<'_, std::sync::Arc<ProjectsService>>,
    request: BindKitRequest,
) -> Result<Project, String> {
    state.bind_delivery_kit(&request).map_err(public_error)
}
#[tauri::command]
pub(crate) fn projects_dependency_snapshot(
    state: State<'_, std::sync::Arc<ProjectsService>>,
    project_id: String,
) -> Result<ProjectDependencySnapshot, String> {
    state.dependency_snapshot(&project_id).map_err(public_error)
}

#[tauri::command]
pub(crate) fn projects_prepare_codex(
    state: State<'_, std::sync::Arc<ProjectsService>>,
    request: ProjectMutation,
) -> Result<ProjectContext, String> {
    state.prepare_codex(&request).map_err(public_error)
}
